conn_mgr misc improvements

glarsennordic opened the pull request:
Miscellaneous improvements to conn_mgr based on feedback over the last few months

glarsennordic committed:
net: conn_mgr: Create separate connectivity implementation API.

Splits many definitions from conn_mgr_connectivity.h off into their own
header, conn_mgr_connectivity_impl.h

conn_mgr_connectivity.h now focuses solely on user/application-facing
APIs, and conn_mgr_connectivity_impl now contains all functions, utils,
definitions intended for use by connectivity implementations, including
conn_mgr_if_get_binding, which is now available for use by connectivity
implementations.

This lays the foundation for allowing connectivity implementations
easier and safer access to their internal state through internal APIs.

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>

glarsennordic committed:
net: conn_mgr: binding lock/unlock

Adds conn_mgr_binding_lock/conn_mgr_binding_unlock, which connectivity
implementations can call to lock or unlock the passed in connectivity
binding.

This allows connectivity implementations to safely make direct accesses
to binding data from outside conn_mgr_conn_api callbacks, and should
allow cleaner, more robust connectivity implementations to be written.

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>

glarsennordic commented on include/zephyr/net/conn_mgr_connectivity_impl.h:
```diff
@@ -252,6 +252,37 @@ static inline struct conn_mgr_conn_binding *conn_mgr_if_get_binding(struct net_i
+ * @param binding
+ * @return int
+ */
+static inline int conn_mgr_binding_lock(struct conn_mgr_conn_binding *binding)
```
would take/give be better names than lock/unlock?

glarsennordic replied:
Reconsider integer return value

glarsennordic committed:
net: conn_mgr: binding get/set flag

Add a helper allowing connectivity implementations to more easily access
connectivity flags on connectivity bindings.

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>

glarsennordic commented on subsys/net/conn_mgr/Kconfig:
```diff
@@ -29,12 +29,20 @@ config NET_CONNECTION_MANAGER_STACK_SIZE
+
+if NET_CONNECTION_MANAGER_AUTO_INIT
+
+config NET_CONNECTION_MANAGER_AUTO_INIT_LEVEL
```
INIT_LEVEL is wrong, it should be INIT_PRIORITY

glarsennordic commented on subsys/net/conn_mgr/conn_mgr.c:
```diff
@@ -334,8 +347,17 @@ static int conn_mgr_init(void)
 }
 
-SYS_INIT(conn_mgr_init, APPLICATION, CONFIG_NET_CONNECTION_MANAGER_PRIORITY);
+SYS_INIT(conn_mgr_auto_init, APPLICATION, CONFIG_NET_CONNECTION_MANAGER_AUTO_INIT_LEVEL);
```
Rather than make the `SYS_INIT` call optional, let's have the `conn_mgr_thread` fully handle initialization, and pend an "all clear" signal.

That way it isn't possible for applications to accidentially initialize `conn_mgr` ***sooner*** than it normally would.

rlubos replied:
Do I understand it correctly, that `SYS_INIT()` would spawn the thread (unconditionally) which would then block on a semaphore for instance, till `conn_mgr_init();` is called?

glarsennordic replied:
Yes, precisely -- although I have decided to skip this feature for this PR

rlubos commented on include/zephyr/net/conn_mgr_connectivity_impl.h:
```diff
@@ -252,6 +252,39 @@ static inline struct conn_mgr_conn_binding *conn_mgr_if_get_binding(struct net_i
+ *
+ * @param binding - Binding to claim
+ */
+static inline void conn_mgr_binding_take(struct conn_mgr_conn_binding *binding)
```
Could be me, but I'd find `conn_mgr_binding_lock()/unlock()` a bit more intuitive.

rlubos commented on include/zephyr/net/conn_mgr_connectivity_impl.h:
```diff
@@ -252,6 +252,39 @@ static inline struct conn_mgr_conn_binding *conn_mgr_if_get_binding(struct net_i
 }
 
+/**
+ * @brief Claims the passed-in binding, making it save to access.
```
"**safe** to access"?

glarsennordic replied:
Oh my, good catch, thanks

glarsennordic committed:
net: conn_mgr: Rename connectivity monitor

The connectivity monitoring subfeature of conn_mgr is currently also
named conn_mgr, which is confusing.

This commit renames it to conn_mgr_monitor, or conn_mgr_mon for short,
for clarity.

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>

glarsennordic commented:
If you are tagged for review and a bit confused, it's because this PR renames a header file and a couple KConfig arguments that get used in a decent number of samples. You can focus on just the last commit if that's the case.

nordicjm commented:
The last commit changes a function name and a filename name, seems like this needs to go through the stable API change process (assuming it is part of networking, which is stable)?

rlubos commented:
> The last commit changes a function name and a filename name, seems like this needs to go through the stable API change process (assuming it is part of networking, which is stable)?

Connection manager is still experimental: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/conn_mgr/Kconfig#L5

nordicjm reviewed (dismissed):
OK for MCUmgr parts

nordicjm commented:
> > The last commit changes a function name and a filename name, seems like this needs to go through the stable API change process (assuming it is part of networking, which is stable)?
> 
> Connection manager is still experimental: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/conn_mgr/Kconfig#L5

Can this be added to the API table? https://docs.zephyrproject.org/latest/develop/api/overview.html

rlubos reviewed (commented):
Looks that maintainers file needs an update as well. Otherwise looks good, but please update https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/develop/api/overview.rst with and entry for Connection Manager so that there's no confusion about the API status.

glarsennordic commented on include/zephyr/net/conn_mgr_connectivity_impl.h:
```diff
@@ -0,0 +1,346 @@
+/**
+ * @brief Check the value of the specified connectivity flag for the provided binding
+ *
+ * Can be used from any thread or callback without calling conn_mgr_binding_take.
```
take->lock (and vice versa)

glarsennordic commented:
@rlubos I can't use `:ref:` at the moment because conn_mgr still doesn't have docs, so I am just using an inline literal for now instead

glarsennordic committed:
net: conn_mgr: Mark as experimental in the API table

Mark conn_mgr as experimental in the API table, since it is currently
experimental.

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>
